`Activator.CreateInstance` supports ByRefLike types

AaronRobinsonMSFT opened the pull request:
This represents an update of `Activator.CreateInstance` to naturally support byreflike generics.
It is modeled off of the current native AOT implementation.

This now works for all runtimes - CoreCLR, mono and native AOT. Native AOT was a no-op as its implementation was ByRefLike ready.

Contributes to https://github.com/dotnet/runtime/issues/65112

AaronRobinsonMSFT committed:
Activator.CreateInstance similar to NAOT

This represents an update of Activator.CreateInstance
to naturally support byreflike generics. It is modeled
off of the current native AOT implementation.

AaronRobinsonMSFT committed:
Mono implementation now supports shared path with CoreCLR.

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into createinstance_update

AaronRobinsonMSFT committed:
Permit ref structs for Activator.CreateInstance and Unsafe.As<F,T>().
Add test

AaronRobinsonMSFT commented:
/cc @jkotas @stephentoub @agocke @AlekseyTs @jaredpar

lambdageek commented on src/mono/mono/metadata/icall.c:
```diff
@@ -4246,6 +4246,17 @@ ves_icall_RuntimeType_GetPropertiesByName_native (MonoQCallTypeHandle type_handl
+	gpointer ctor_fptr;
+	MonoReflectionMethod* ctor = *(MonoReflectionMethod**)ref_ctor_info;
+	ctor_fptr = mono_compile_method_checked (ctor->method, error);
+	if (!is_ok (error))
+		return NULL;
```
this will cause a managed exception to be thrown if there's an `error` (for example if we trigger assembly loading and the assembly isn't found, or if the method uses an invalid class we'll throw `BadImageFormatException` or something)  is that what you want? or do you just want to return null and throw a generic `ExecutionEngineException` up in managed?

AaronRobinsonMSFT replied:
I want the exception to be thrown if it exists. If an exception isn't thrown, but `NULL` is returned, we will throw an `ExecutionEngineException`.

jaredpar commented on src/libraries/System.Private.CoreLib/src/System/Activator.RuntimeType.cs:
```diff
@@ -136,8 +136,20 @@ public static partial class Activator
+            if (!rtType.IsValueType)
+            {
+                object o = rtType.CreateInstanceOfT()!;
+                return Unsafe.As<object, T>(ref o);
```
Did you consider adding a comment specifying why this specifically is safe to use here? Left undocumented customers, internal or external, may think it's valid ot down cast arbitrary `object` to `ref struct`.

AaronRobinsonMSFT replied:
I love that reasoning for a comment. Thank you.

AaronRobinsonMSFT committed:
Comment why Unsafe.As<> is safe.
Wrap invocation in TargetInvocationException.

lambdageek commented:
Mono crashes look related. I felt pretty certain just compiling the ctor would return something with the correct calling convention, but perhaps I need to verify that

lambdageek committed:
Use MethodHandle.GetFunctionPointer to call the value constructor

On the Mono interpreter, the ldftn opcode for a managed method returns a MonoFtnDesc*, not
a native code pointer, and the calli opcode expects a MonoFtnDesc*.

The MethodHandle.GetFunctionPointer() function is an interpreter
intrinsic that returns such a MonoFtnDesc*

lambdageek commented:
@AaronRobinsonMSFT please cherrypick https://github.com/lambdageek/runtime/commit/c4067c404570d838c4c32bb62728387b9ebb0b6c

The issue was that when using the interpreter `ldftn`/`calli` operate on a `MonoFtnDesc*` not a native code pointer (which is what `mono_compile_method_checked` returns).  `RuntimeMethodHandle.GetFunctionPointer` is an interpreter intrinsic that gets replaced by an `ldftn`.

---

there's a second crash in `System.Runtime.Tests` on `Build linux-x64 Debug Mono_MiniJIT_LibrariesTests` that looks like a crash while creating a stack trace after an xunit test failure.  My patch apparently fixes that, too, although I'm unable to explain why - my best guess is that in some generic sharing situation we weren't passing the generic sharing context correctly (`get_ftnptr_for_method` in mini-runtime.c does a bit more work than just `mono_compile_method_checked`)

AaronRobinsonMSFT commented on src/mono/System.Private.CoreLib/src/System/RuntimeType.Mono.cs:
```diff
@@ -1615,6 +1615,34 @@ private void CreateInstanceCheckThis()
+
+            // Important: when using the interpreter, GetFunctionPointer is an intrinsic that
+            // returns a function descriptor suitable for casting to a managed function pointer.
+            // Other ways of obtraining a function pointer might not work.
```
```suggestion
            // Other ways of obtaining a function pointer might not work.
```

AaronRobinsonMSFT committed:
Update src/mono/System.Private.CoreLib/src/System/RuntimeType.Mono.cs

AaronRobinsonMSFT commented:
> @AaronRobinsonMSFT please cherrypick https://github.com/lambdageek/runtime/commit/c4067c404570d838c4c32bb62728387b9ebb0b6c

Thanks @lambdageek. I've applied the changes. Let's see if the CI is happy now.

lambdageek commented on src/mono/System.Private.CoreLib/src/System/RuntimeType.Mono.cs:
```diff
@@ -1615,6 +1615,34 @@ private void CreateInstanceCheckThis()
+
+            RuntimeConstructorInfo? ctor = GetDefaultConstructor();
+            if (ctor == null)
+                return;
```
`CreateInstanceT_StructWithPrivateDefaultConstructor_ThrowsMissingMethodException` is failing on Mono, perhaps due to this being a `return` instead of a `throw`

AaronRobinsonMSFT replied:
It is a bit more complicated than that, but thanks for point out the issue. If the `ctor` is `null`, that means it isn't defined, which turns into a nop. However, if the `ctor` is defined, it must be `public`. That second check is what I am missing.

AaronRobinsonMSFT committed:
Missed public ctor check.

AaronRobinsonMSFT committed:
Keep API names consistent across runtimes.

AaronRobinsonMSFT commented on src/mono/System.Private.CoreLib/src/System/RuntimeType.Mono.cs:
```diff
@@ -37,7 +37,7 @@ internal enum TypeNameFormatFlags
     }
 
-    internal partial class RuntimeType
+    internal unsafe partial class RuntimeType
```
```suggestion
    internal partial class RuntimeType
```
